[MRG] Validate MD5 checksum of downloaded ARFF in fetch_openml

chadykamar opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Fixes #11816 


#### What does this implement/fix? Explain your changes.
Compares the MD5 hash of the downloaded ARFF with the checksum provided in the OpenML data set description. Issues a warning if they are not equal. 

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

chadykamar committed:
Validate MD5 checksum

jnothman commented:
Why closed?

jnothman commented:
Please don't close just because there are test failures

chadykamar committed:
Merge changes

chadykamar committed:
Remove data desc false checksum files

chadykamar committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into md5-checksum

chadykamar committed:
Add comment

chadykamar commented:
Sorry didn't mean to do that exactly

chadykamar committed:
Fix build error related to tmpdir fixture

chadykamar committed:
Fix json TypeError in Python 2 build

rth commented on sklearn/datasets/openml.py:
```diff
@@ -45,13 +45,32 @@ def _open_openml_url(openml_path, data_home):
+                warn('Data set file hash {} does not match the checksum {}.'
+                     .format(md5_hash, md5_checksum))
+
+        fp = tempfile.TemporaryFile()
```
It would be preferable to avoid creating a temporary file. Maybe using `BytesIO` ?

rth commented on sklearn/datasets/openml.py:
```diff
@@ -63,7 +82,15 @@ def _open_openml_url(openml_path, data_home):
+                        warn('Data set file hash {} does not match the'
+                             ' checksum {}.'.format(md5_hash, md5_checksum))
+
+                fdst.write(content)
```
This is not strictly equivalent. In `copyfileobj`,

> by default the data is read in chunks to avoid uncontrolled memory consumption

I'm not fully sure whether hashing by chunks (see e.g. these [SO answers](https://stackoverflow.com/questions/1131220/get-md5-hash-of-big-files-in-python)) is worthwhile here

chadykamar replied:
I'll do that, it might be necessary for very large data files.

rth commented on sklearn/datasets/tests/test_openml.py:
```diff
@@ -527,3 +511,94 @@ def test_fetch_openml_raises_illegal_argument():
+    # Capture all warnings
+    with pytest.warns(None) as records:
+        fetch_openml(data_id=data_id, data_home=str(tmpdir), cache=cache)
+        # assert no warnings
```
Above 2 comments are unecessary IMO

rth reviewed (commented):
Thanks @chadykamar , a few comments from a partial review are below.

chadykamar committed:
Hash and write by chunks

- Replace temp file with BytesIO when  not caching.
- Hashing and writing by chunks
- Removed unecessary comments in test_openml.py

chadykamar committed:
Remove unnecessary copy

chadykamar committed:
Fix inadvertently pass test

chadykamar committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into md5-checksum

rth commented on sklearn/datasets/tests/test_openml.py:
```diff
@@ -127,11 +127,6 @@ def _fetch_dataset_from_openml(data_id, data_name, data_version,
-    # test return_X_y option
-    fetch_func = partial(fetch_openml, data_id=data_id, cache=False,
-                         target_column=target_column)
-    check_return_X_y(data_by_id, fetch_func)
```
Why was this removed?

chadykamar replied:
I made a mistake when merging.

rth commented on sklearn/datasets/tests/test_openml.py:
```diff
@@ -214,23 +209,11 @@ def test_fetch_openml_iris(monkeypatch):
-        UserWarning,
-        "Multiple active versions of the dataset matching the name"
-        " iris exist. Versions may be fundamentally different, "
-        "returning version 1.",
```
Same comment: we still want to check this warning message, unless I am missing something.

rth reviewed (approved):
Generally this LGTM, apart for the two comments below.

(Maybe it was a merge that went wrong?)

chadykamar committed:
Correct merging error

rth commented on sklearn/datasets/tests/test_openml.py:
```diff
@@ -215,21 +217,21 @@ def test_fetch_openml_iris(monkeypatch):
+            'expect_sparse': False,
+            'expected_data_dtype': np.float64,
+            'expected_target_dtype': object,
+            'compare_default_target': True}
```
In all the above, changes are due to fixing the merge. We don't want to see any of these changes in this diff because it would complicate e.g. git blame. So the easiest to remove these changes would be,
 - either manually with `git diff master...HEAD` (or `HEAD...master` I never can remember) and editing the file until none of it shows.
 - or copying this file somewhere, reverting to the version on master `git checkout master -- <path_to_this_file>` than adding back your changes.
Don't worry about individual commits as in the end everything will be squashed into a single one, so only the total diff matter. Apart for this this looks good.

jnothman commented on sklearn/datasets/openml.py:
```diff
@@ -399,6 +436,10 @@ def fetch_openml(name=None, version='active', data_id=None, data_home=None,
 
+    verify_checksum : boolean, default=True
+        Whether or not to validate that the dataset file's MD5 hash matches the
+        data set description's expected checksum.
```
Maybe note that "Verification occurs when fetching the data, so if cache=True, the data will be presumed valid on subsequent calls."

chadykamar committed:
Merge mishaps

jnothman commented on sklearn/datasets/openml.py:
```diff
@@ -63,7 +86,19 @@ def _open_openml_url(openml_path, data_home):
+                        fdst.write(block)
+
+                    if md5_checksum != md5.hexdigest():
+                        warn('Data set file hash {} does not match the '
```
Why do we want a warning rather than an exception?

chadykamar replied:
I thought an exception would be raised when decoding in `_arff`, but it might be better to fail earlier. Which exception would you raise instead?

jnothman replied:
If it's incomplete but stops at a line break, there will be no exception. We should raise one here

lesteve commented on sklearn/datasets/openml.py:
```diff
@@ -45,13 +46,35 @@ def _open_openml_url(openml_path, data_home):
+        if md5_checksum is None:
+            return response
+
+        stream = io.BytesIO()
```
there seems to be some duplication between here and a few lines below. Could you factor that out in a function or something?

chadykamar committed:
Change warning into exeception

chadykamar committed:
Add doc detail for verify_checksum param

rth commented:
Could you please resolve conflicts ? There are also a few CI failures..

rth commented:
Tried to resolve conflicts, but it's not trivial. I think it would probably easier to add these changes on top of upstream master manually.

deeplook commented on sklearn/datasets/openml.py:
```diff
@@ -72,6 +90,20 @@ def _open_openml_url(openml_path, data_home):
+    if md5_checksum != md5.hexdigest():
+        msg = 'Data set file hash {} does not match the checksum {}.'
+        msg = msg.format(md5.hexdigest(), md5_checksum)
+        raise Exception(msg)
```
This does more than just *check* the MD5 hash of a file it also writes to a file stream. So it is at least named misleadingly.

deeplook commented:
@chadykamar Are you still on this? I would give it a try, even restarting on top of master, but I don't want you to loose any credit for your changes so far.

deeplook commented:
Following @adrinjalali's advice I'll give it a try and start from scratch now.

adrinjalali commented:
Closing as #14800 fixes the issue.
